### ISSUE TITLE
OpenTSDB leaking Sockets?

### ISSUE THREAD

**Issue by kylebrandt** [OpenTSDB/opentsdb#304]
I've hit too many open files, digging into this:
```
[root@ny-devtsdb04 fd]# ps aux | grep opentsdb.conf
root      2076  3.8  1.5 5489036 260496 pts/0  Sl   Mar21  61:23 java -enableassertions -enablesystemassertions -classpath /usr/local/share/opentsdb/asynchbase-1.4.1.jar:/usr/local/share/opentsdb/guava-13.0.1.jar:/usr/local/share/opentsdb/jackson-annotations-2.2.3.jar:/usr/local/share/opentsdb/jackson-core-2.2.3.jar:/usr/local/share/opentsdb/jackson-databind-2.2.3.jar:/usr/local/share/opentsdb/log4j-over-slf4j-1.7.2.jar:/usr/local/share/opentsdb/logback-classic-1.0.9.jar:/usr/local/share/opentsdb/logback-core-1.0.9.jar:/usr/local/share/opentsdb/netty-3.6.2.Final.jar:/usr/local/share/opentsdb/slf4j-api-1.7.2.jar:/usr/local/share/opentsdb/suasync-1.4.0.jar:/usr/local/share/opentsdb/tsdb-2.0.0.jar:/usr/local/share/opentsdb/zookeeper-3.3.6.jar:/usr/local/share/opentsdb net.opentsdb.tools.TSDMain --port=4242 --staticroot=build/staticroot --cachedir=/tmp/tsd --zkquorum=ny-devtsdb01,ny-devtsdb03,ny-devtsdb04 --config /root/dev/opentsdb.conf
root     23362  0.0  0.0 103256   872 pts/0    S+   17:32   0:00 grep opentsdb.conf
[root@ny-devtsdb04 fd]# ls -l /proc/2076/fd | grep socket | wc -l
4049
[root@ny-devtsdb04 fd]# netstat -ap | grep 2076 | wc -l
6
[root@ny-devtsdb04 fd]# lsof -p 2076 | grep sock | head
java    2076 root   17u  sock                0,6      0t0 13372319 can't identify protocol
java    2076 root   49u  sock                0,6      0t0 13394689 can't identify protocol
java    2076 root   50u  sock                0,6      0t0 13394637 can't identify protocol
java    2076 root   51u  sock                0,6      0t0 13395575 can't identify protocol
java    2076 root   52u  unix 0xffff880437c5a0c0      0t0 12977998 socket
java    2076 root   53u  sock                0,6      0t0 12977996 can't identify protocol
java    2076 root   54u  sock                0,6      0t0 13394738 can't identify protocol
java    2076 root   56u  sock                0,6      0t0 13395172 can't identify protocol
java    2076 root   57u  sock                0,6      0t0 13394785 can't identify protocol
java    2076 root   58u  sock                0,6      0t0 13376378 can't identify protocol
```

I was doing some maintenance the other day on my hbase cluster and restarting region servers, so maybe when stuff "disappears" in the cluster they don't get closed properly? 

(Still need to update async base, maybe that is part of it..?)

**Comment by jmangs** [comment]
Looks like for Netty 3.5.9 and later, the `IdleStateEvent` is now handed to the I/O worker thread. Was changed here [#641](https://github.com/netty/netty/issues/641) - I'll try taking a look at this again sometime over the weekend since it's just a matter of making sure the I/O threads handle the event now.....

**Comment by kylebrandt** [comment]
For now I'm seeing if I still have this problem with async_io disabled as tkrajca said. I don't know really know what is going on, other than the following happens to me now:
- Bosun throws "use of closed network connection" errors 
- OpenTSDB's CLOSE_WAIT sockets increase when this happens. So do the number of FDs (sockets I presume) for both Bosun and OpenTSDB.
- CPU on OpenTSDB goes up when this happens
- Connection Manager for TSD throws some exceptions
- GC Spiked on one of my region servers during this

![image](https://cloud.githubusercontent.com/assets/1692624/6079683/68e0cff4-add6-11e4-9c71-8a6917e3a197.png)


![image](https://cloud.githubusercontent.com/assets/1692624/6079680/5fd3164c-add6-11e4-8e8f-c56edcecd459.png)

![image](https://cloud.githubusercontent.com/assets/1692624/6079687/7e63b8be-add6-11e4-92a6-24bddd2f1b55.png)

![image](https://cloud.githubusercontent.com/assets/1692624/6079692/8c308d32-add6-11e4-8f1b-6358705634a1.png)

**Comment by jmangs** [comment]
@kylebrandt, could you see if the changes in #439 fix the issues with sockets getting stuck in CLOSE_WAIT again? Like I mentioned above, in Netty 3.5.9 and later, I/O threads handle `IdleStateEvents` now so what was closing sockets previously no longer functioned (because it was written with an older version of Netty in mind).

**Comment by kylebrandt** [comment]
@jmangs Yes but probably not until sometime next week (and this problem happens roughly once a week) so might need to be a bit patient :-/

**Comment by haridsv** [comment]
We just hit a very similar problem in many of our production clusters. Some kind of glitch caused NSRE exceptions for hbase:meta table and this caused the OpenTSDB servers to go into a sort of a limbo and start leaking sockets and finally maxed out the limits.

I was able to reproduce a very similar scenario in a test cluster by force-closing hbase:meta for a few minutes while there is significant load on the server. What I noticed is that all the worker threads started waiting indefinitely on hbase requests and this made them no longer consume their task queues. Since the boss thread is still functional, it is able to accept connection requests and add them to the task queue, which causes the sockets to leak and ultimately reach the limit to cause accept itself to fail (in theory).

Now I don't know why the worker threads are indefinitely waiting. From the code, it looks like they are supposed to be woken up by the async hbase threads, and these threads seem to be functional, so perhaps there is an exception handler missing that caused them to miss some callbacks? I don't where to look for this code, so I am clueless on this. But I am sure I can reproduce this easily and can even do a bit of debugging (limited to jdb). 

We are still using version 2.0.0 and so might not have some of the fixes discussed earlier in this thread, but wondering if this could be a different issue or the same. I will try to get the latest version to work in our environment, but is there something I can do in the test environment to further troubleshoot and narrow down this issue?

**Comment by jmangs** [comment]
I've seen this as well when we've done rolling restarts of HBase; in your case I think the workers are all stuck in waiting for HBase to come back; usually it will never full recover and you'll need to restart the TSD. We've had this happen in 2.1.0 as well and it also affects utilities like `fsck` - usually the logs are spammed with warnings about a region server being offline or the root region being offline and that the cache has been invalidated.

This isn't the same issue above though, previously TSDs would also let sockets linger forever because there was no code on the Netty side to close them after being idle. In the case with NSREs, the workers just stop processing requests and all of the requests get backed up forever until you restart the affected TSD - if you don't then it just keeps opening sockets that are all waiting on HBase to come back. Not sure if there's an open issue for this yet though. As for recreating it, doing rolling restarts of region servers will usually trigger your problem at least in my experience (and it affects 2.1.0).

**Comment by haridsv** [comment]
Thanks @jmangs for the response! In a rolling restart, do the NSRE's have to last for long time for this scenario to occur? We also do rolling restarts but move the regions temporarily to other region servers (and then move them back), so any NSRE's generated in those two short windows should be similar to an NSRE generated during a split and so should be handled gracefully, right? However, I will work with our ops guys to see if this can be observed in other clusters that had a rolling restart.

**Comment by jmangs** [comment]
In my experience it seems that once enough NSREs are thrown and RPCs are delayed it will have trouble coming back fully. I'm fairly certain there's some sort of race condition in the code since I've had cases where NSREs didn't cause your issue to occur. I can't give you an exact number but to work around this we usually coordinate rolling restarts to make sure that we watch our TSDs to see if they're choking. Usually you will see a spike in HBase latency in the statistics (and the API endpoints fail to return a response, you just wait forever)

**Comment by haridsv** [comment]
Ok, makes sense, and I agree there has to be a race condition or a missing exception handling that causes the worker threads to wait for ever.

Could someone point me to the are of code that is supposed to listen to these completion callbacks and notify the waiting workers? I am willing to spend some time to debug this issue.

**Comment by jmangs** [comment]
I'm not sure but I'd check out the callback handlers for errors. I think the last time I fixed an issue like this in OpenTSDB I took a thread dump of the JVM using `jstack` and looked at where all of the threads were stuck. Probably the easiest way to find where it's getting stuck since the async callbacks are hard to follow in the code.

**Comment by haridsv** [comment]
I already have a thread (and heap dump) and all that the worker thread stacks simply show them waiting to be woken up.

I am not sure what the reasoning is behind not using a wait timeout, but it makes perfect sense to have a configured timeout after which the requests would be marked as failed and free the worker threads to continue with the rest of the queue (assuming that the async hbase does recover from cluster availability issues). This doesn't of course fix the root cause, but can serve as a workaround. I can make such a change and reproduce this issue to see if it can recover.

**Comment by haridsv** [comment]
@jmangs I noticed your pull request #290 and left a couple of comments, could you take a look?

**Comment by IzakMarais** [comment]
We have also encountered this. After an HBase NSRE OpenTSDB 2.1.0 leaks connections until it becomes unresponsive and requires a restart. 

[Here is a Grafana snapshot](https://snapshot.raintank.io/dashboard/snapshot/TSxQ8KqrYETP0MyLlpuIatbZArdN04qq) confirming the behaviour.

**Comment by gansvv** [comment]
I'm running a test to verifying stability of OpenTSDB. I see a large number of CLOSE_WAIT connections after a HBASE NSRE, and OpenTSDB becoming unresponsive soon after.
```
$ sudo netstat -anp | fgrep 4242 |  awk '{print $6}' | sort | uniq -c
   1208 CLOSE_WAIT
    601 ESTABLISHED
      1 FIN_WAIT2
      1 LISTEN
```
Are there better known mitigation steps for this issue other that restarting the process after each HBase NSRE? Is the root cause and fix-required known? I'm willing to help. Thanks.

**Comment by lordang** [comment]
Are there any progress this issue? We have read and write node separately and we see this issue only on read tsd node. And both are same environments. We have upgraded Netty jar 3.9.5 just in case.

**Comment by IzakMarais** [comment]
> We have read and write node separately and we see this issue only on read tsd node.
We have the same behaviour. 
(I have written a service that monitors and kills the reader tsd if its open file handles exceed 1000. We use systemd to manage it, so it will be automatically restarted).

**Comment by johann8384** [comment]
@manolama do you know what we need to do to address this?

**Comment by lordang** [comment]
@IzakMarais Maybe I should do that. thanks.

**Comment by haridsv** [comment]
We have put in a temporary workaround of using a timeout on all the joins,
something like this:

-      return metricNameAsync().joinUninterruptibly();
+      return
metricNameAsync().joinUninterruptibly(Const.ABS_ASYNCHBASE_JOIN_TIMEOUT_MS);

Where the const is defined as:

+  public static final long ABS_ASYNCHBASE_JOIN_TIMEOUT_MS = 10*60*1000;

This is only a workaround, as it doesn't prevent this condition from
reaching, but it helps the service in recovering from this situation by
itself.

**Comment by manolama** [comment]
Hmm, I swear we had something to fix this in the OSS code but I know I haven't seen it in a while here. Let me go through and find what's in the connection path that may have changed.
